libparsec/crates/client/src/workspace/merge.rs

```diff
@@ -1,150 +1,315 @@
-            // - the fixtures and server data binder system used in the tests
-            //   makes it much more likely
-            let is_speculative_fn = $is_speculative_fn;
-            if $remote.author == *$local_author && !is_speculative_fn($local) {
-                let mut new_local = $local.to_owned();
-                new_local.base = $remote;
-                return Some(new_local);
-            }
+    // Both the remote and the local have changed
 
-            // The remote has been updated by some other device
-
-            // Solve the folder conflict
-            let new_children = merge_children(&$local.base.children, &$local.children, &$remote.children);
-
-            // Children merge can end up with nothing to sync.
-            //
-            // This is typically the case when we sync for the first time a workspace
-            // shared with us that we didn't modify:
-            // - the workspace manifest is a speculative placeholder (with arbitrary update&create dates)
-            // - on sync the update date is different than in the remote, so a merge occurs
-            // - given we didn't modify the workspace, the children merge is trivial
-            // So without this check each each user we share the workspace with would
-            // sync a new workspace manifest version with only it updated date changing :/
-            //
-            // Another case where this happen:
-            // - we have local change on our workspace manifest for removing an entry
-            // - we rely on a base workspace manifest in version N
-            // - remote workspace manifest is in version N+1 and already integrate the removal
-            //
-            // /!\ Extra attention should be paid here if we want to add new fields
-            // /!\ with their own sync logic, as this optimization may shadow them!
-
-            if new_children == $remote.children {
-                Some(<$local_xxx_manifest_ty>::from_remote($remote, None))
-            } else {
-                let mut local_from_remote = <$local_xxx_manifest_ty>::from_remote($remote, None);
-                local_from_remote.children = new_children;
-                local_from_remote.updated = $timestamp;
-                local_from_remote.need_sync = true;
-                Some(local_from_remote)
+    // 3) The remote changes are ours (our current local changes occurs while
+    // we were uploading previous local changes that became the remote changes),
+    // simply acknowledge the remote changes and keep our local changes
+    if remote_manifest.author == *local_author {
+        let mut new_local = local_manifest.to_owned();
+        new_local.base = remote_manifest;
+        return MergeLocalFileManifestOutcome::Merged(new_local);
+    }
+
+    // 4) Merge data and ensure the sync is still needed
+
+    // Destruct local and remote manifests to ensure this code with fail to compile
+    // whenever a new field is introduced.
```

Comment:
I don't quite get why you would want that ? Is it to have a real human person check that the new field does not interact with the merging process, or to adapt the merging process ? 

Maybe annotating LocalFileManifest with [non_exhaustive](https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute) could do the trick

Reply:
Consider the following:

```rust
struct Data {
  updated: DateTime,
}

fn merge(local: Data, remote: Data) -> Data {
   remote.updated = max(local.updated, remote.updated);
   remote
}
```

Then we introduce a new `created` field

```rust
struct Data {
  created: DateTime,
  updated: DateTime,
}
```

The code will compile fine event if we haven't updated `merge` (typically we should have added a line like `remote.created = min(local.created, remote.created)`).

This is an issue given this `merge` function is easy to forget to update:
- it is in a different file
- its tests will most likely just ignore the new field

On the other hand, if the `merge` function is implemented with exhaustive de-structuring, it will fail to compile any time a new field is added. Hence forcing the developer to have a look at this part of the code.

This is a pattern we also use in the [CRC code of the testbed](https://github.com/Scille/parsec-cloud/blob/30f3d2d238759bce33eb3c104055b542c268d4d4/libparsec/crates/testbed/src/template/crc_hash.rs#L316-L322). Nobody knows about this piece of code and nobody cares... but if a structure changes and the CRC starts being wrong, we will have very hard time understanding why the tests start becoming randomly wrong (as client and testbed server code may end up with different data with the same CRC...)

Regarding `non_exhaustive`, its point is to forbid exhausting destructuring, which is the opposite of what we want here ;-)

Reply:
Thanks, so we could put that behind a debug_assert to make sure that it won't be present during runtime ?